src/node_crypto.cc

```diff
@@ -2563,6 +2564,841 @@ static bool IsSupportedAuthenticatedMode(int mode) {
+      .ToLocalChecked();
+  KeyObject* key = Unwrap<KeyObject>(obj);
+  CHECK(key);
+  if (key_type == kKeyTypePublic)
+    key->InitPublic(pkey);
+  else
+    key->InitPrivate(pkey);
+  return obj;
+}
+
+ManagedEVPPKey KeyObject::GetAsymmetricKey() const {
+  CHECK_NE(key_type_, kKeyTypeSecret);
+  return this->asymmetric_key_;
+}
+
+const char* KeyObject::GetSymmetricKey() const {
+  CHECK_EQ(key_type_, kKeyTypeSecret);
+  return this->symmetric_key_.get();
+}
+
+size_t KeyObject::GetSymmetricKeySize() const {
+  CHECK_EQ(key_type_, kKeyTypeSecret);
+  return this->symmetric_key_len_;
+}
+
+void KeyObject::New(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args.IsConstructCall());
+  CHECK(args[0]->IsInt32());
+  KeyType key_type = static_cast<KeyType>(args[0].As<Uint32>()->Value());
+  Environment* env = Environment::GetCurrent(args);
+  new KeyObject(env, args.This(), key_type);
+}
+
+KeyType KeyObject::GetKeyType() const {
+  return this->key_type_;
+}
+
+void KeyObject::Init(const FunctionCallbackInfo<Value>& args) {
+  KeyObject* key;
+  ASSIGN_OR_RETURN_UNWRAP(&key, args.Holder());
+
+  unsigned int offset;
+  ManagedEVPPKey pkey;
+
+  switch (key->key_type_) {
+  case kKeyTypeSecret:
+    CHECK_EQ(args.Length(), 1);
+    key->InitSecret(Buffer::Data(args[0]), Buffer::Length(args[0]));
+    break;
+  case kKeyTypePublic:
+    CHECK_EQ(args.Length(), 3);
+
+    offset = 0;
+    pkey = GetPublicKeyFromJS(args, &offset, false, false);
+    if (!pkey)
+      return;
+    key->InitPublic(pkey);
+    break;
+  case kKeyTypePrivate:
+    CHECK_EQ(args.Length(), 4);
+
+    offset = 0;
+    pkey = GetPrivateKeyFromJS(args, &offset, false);
+    if (!pkey)
+      return;
+    key->InitPrivate(pkey);
+    break;
+  default:
+    CHECK(false);
+  }
+}
+
+void KeyObject::InitSecret(const char* key, size_t key_len) {
+  CHECK_EQ(this->key_type_, kKeyTypeSecret);
+
+  char* mem = CHECKED_OPENSSL_malloc<char>(key_len);
+  memcpy(mem, key, key_len);
+  this->symmetric_key_ = std::unique_ptr<char, std::function<void(char*)>>(mem,
+      [key_len](char* p) {
+        OPENSSL_clear_free(p, key_len);
+      });
+  this->symmetric_key_len_ = key_len;
+}
+
+void KeyObject::InitPublic(const ManagedEVPPKey& pkey) {
+  CHECK_EQ(this->key_type_, kKeyTypePublic);
+  CHECK(pkey);
+  this->asymmetric_key_ = pkey;
+}
+
+void KeyObject::InitPrivate(const ManagedEVPPKey& pkey) {
+  CHECK_EQ(this->key_type_, kKeyTypePrivate);
+  CHECK(pkey);
+  this->asymmetric_key_ = pkey;
+}
+
+Local<String> KeyObject::GetAsymmetricKeyType() const {
+  CHECK_NE(this->key_type_, kKeyTypeSecret);
+  const char* name;
+  switch (EVP_PKEY_id(this->asymmetric_key_.get())) {
+  case EVP_PKEY_RSA:
+    name = "rsa";
+    break;
+  case EVP_PKEY_DSA:
+    name = "dsa";
+    break;
+  case EVP_PKEY_EC:
+    name = "ec";
+    break;
+  default:
+    printf("%d\n", EVP_PKEY_id(this->asymmetric_key_.get()));
+    CHECK(false);
+  }
+  return OneByteString(env()->isolate(), name);
```

Comment:
It might be worth using persistent string values (i.e. the list in `env.h`) here for performance?